rllib/models/specs/specs_dict.py

```diff
@@ -127,4 +138,151 @@ def validate(
+            `self` as the first argument, `input_data` as the second argument and any
+            other keyword argument thereafter. It should return a single object.
+        input_spec: `self` should have an instance method that is named input_spec and
+            returns the `ModelSpec`, `TensorSpec`, or simply the `Type` that the
+            `input_data` should comply with.
+        output_spec: `self` should have an instance method that is named output_spec
+            and returns the spec that the output should comply with.
+        filter: If True, and `input_data` is a nested dict the `input_data` will be
+            filtered by its corresponding spec tree structure and then passed into the
+            implemented function to make sure user is not confounded with unnecessary
+            data.
+        cache: If True, only checks the input/output validation for the first time the
+            instance method is called.
+        input_exact_match: If True, the input data (should be a nested dict) must match
+            the spec exactly. Otherwise, the data is validated as long as it contains
+            at least the elements of the spec, but can contain more entries.
+        output_exact_match: If True, the output data (should be a nested dict) must
+            match the spec exactly. Otherwise, the data is validated as long as it
+            contains at least the elements of the spec, but can contain more entries.
+
+    Returns:
+        A wrapped instance method. In case of `cache=True`, after the first invokation
+        of the decorated method, the intance will have `__checked_specs_cache__`
+        attribute that store which method has been invoked at least once. This is a
+        special attribute can be used for the cache itself. The wrapped class method
+        also has a special attribute `__checked_specs__` that marks the method as
+        decorated.
+    """
+
+    if not input_spec and not output_spec:
+        raise ValueError("At least one of input_spec or output_spec must be provided.")
+
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(self, input_data, **kwargs):
+
+            if cache and not hasattr(self, "__checked_specs_cache__"):
+                self.__checked_specs_cache__ = {}
+
+            def should_validate():
+                return not cache or func.__name__ not in self.__checked_specs_cache__
+
+            def validate(data, spec, exact_match, tag="data"):
+                is_mapping = isinstance(spec, ModelSpec)
+                is_tensor = isinstance(spec, TensorSpec)
+
+                if is_mapping:
+                    if not isinstance(data, Mapping):
+                        raise ValueError(
+                            f"{tag} must be a Mapping, got {type(data).__name__}"
+                        )
+                    data = NestedDict(data)
+
+                if should_validate():
```

Comment:
should you check should_validate() first thing, so we don't waste time building NestedDict, etc?

Reply:
👌 👌